src/main/java/com/redhat/devtools/lsp4ij/features/diagnostics/LSPDiagnosticAnnotator.java

```diff
@@ -66,28 +91,86 @@ public void doApply(@NotNull PsiFile file, Boolean applyAnnotator, @NotNull Anno
             return;
         }
         URI fileUri = LSPIJUtils.toUri(file);
-        Document document = LSPIJUtils.getDocument(file.getVirtualFile());
+        VirtualFile virtualFile = file.getVirtualFile();
+        if (virtualFile == null) {
+            return;
+        }
+        Document document = LSPIJUtils.getDocument(virtualFile);
+        if (document == null) {
+            return;
+        }
 
         // Loop for language server which report diagnostics for the given file
-        var servers = LanguageServiceAccessor.getInstance(file.getProject())
-                .getStartedServers();
+        var servers = LanguageServiceAccessor.getInstance(file.getProject()).getStartedServers();
         for (var ls : servers) {
-            LSPVirtualFileData data = ls.getLSPVirtualFileData(fileUri);
-            if (data != null) {
-                // The file is mapped with the current language server
-                var ds = data.getDiagnosticsForServer();
-                // Loop for LSP diagnostics to transform it to Intellij annotation.
-                for (Diagnostic diagnostic : ds.getDiagnostics()) {
-                    ProgressManager.checkCanceled();
-                    createAnnotation(diagnostic, document, file, ds, holder);
+            // TODO: Is there not a way to check whether the server is valid for this file here?
+            // Force a connection for batch inspection
+            boolean needsDisconnect = forBulkInspection && connectIfNecessary(ls, file);

```

Comment:
In a bulk inspection run, connect the file to the language server if not already connected and note that we did so so that we can disconnect the file when we're finished with it.